### PR TITLE
8333685: Make update-copyright-year script more useful

### PR DESCRIPTION
**Pull request by SoniaZaldana**
Hi all, 

This PR addresses [8333685](https://bugs.openjdk.org/browse/JDK-8333685). 

I have added the following enhancements: 
- Removed uses of `fgrep` and `egrep` with `grep -F` and `grep -E` respectively. 
- Altered default behaviour to limit the processed changesets to those in the current branch and the current year. 
- Enabled an option to modify all changesets in the year if desired (this was the previous default behaviour). 
- Added named parameters and enabled `--help`.
- Removed mercurial support. 
- Fixed a bug where copyright headers that didn't have a comma following the initial year of creation were not getting picked up. For example, [here](https://github.com/openjdk/jdk/blob/512b2b4f141f9a202984150b0427372e1a409a50/src/hotspot/cpu/zero/copy_zero.hpp#L3). 
- Fixed a bug where copyright headers missing the copyright symbol (c) were not getting picked up. Refer to the example above as well. 

Thanks, 
Sonia

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8333685](https://bugs.openjdk.org/browse/JDK-8333685): Make update-copyright-year script more useful (**Enhancement** - P4)

### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19605/head:pull/19605` \
`$ git checkout pull/19605`

Update a local copy of the PR: \
`$ git checkout pull/19605` \
`$ git pull https://git.openjdk.org/jdk.git pull/19605/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19605`

View PR using the GUI difftool: \
`$ git pr show -t 19605`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19605.diff">https://git.openjdk.org/jdk/pull/19605.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19605#issuecomment-2158523570)

### CODE REVIEW

**Comment by tstuefe** [comment]
Thanks for tackling this.

When it will work, it will be very useful, but so far I did not get it to work on my Mac m1.

It identifies the commits correctly that are part of the delta to master, as it should. But it does not modify any copyrights. I tracked this down to the sed's in updateFile not doing anything.

I tried the script on this branch: https://github.com/tstuefe/jdk/tree/do-some-stuff-in-aix and called the script without command line args. It should have found and fixed the Oracle copyrights in osThread_aix.xxx

For reference, this is how the expanded sed arguments look like:

```
s@\(Copyright \((c) \)\?[12][0-9][0-9][0-9],\) [12][0-9][0-9][0-9], Oracle@\1 2024, Oracle@
s@\(Copyright \((c) \)\?[12][0-9][0-9][0-9],\) Oracle@\1 2024, Oracle@
s@\(Copyright \((c) \)\?[12][0-9][0-9][0-9]\) Oracle@\1, 2024, Oracle@
s@Copyright 2024, 2024, Oracle@Copyright 2024, Oracle@
s@Copyright (c) 2024, 2024, Oracle@Copyright (c) 2024, Oracle@
```

**Comment by erikj79** [comment]
> Tested on Linux, there it works.
> 
> Note, IIRC, the original unpatched version also worked on MacOS. So, I think the original sed expressions were probably ok. Something about `\(${copyright_symbol} \)\?` perhabs.

Mac has bsd sed by default which does not handle '?'. You would have to express it as `\{0,1\}` instead to be compatible.

**Comment by pavelrappo** [comment]
Opinion: while it's good to see improvements to the existent script, since JEP 330, we can now conveniently implement a similar script in Java. That'll also automatically take care of OS specifics.

**Comment by tstuefe** [comment]
> Opinion: while it's good to see improvements to the existent script, since JEP 330, we can now conveniently implement a similar script in Java. That'll also automatically take care of OS specifics.

Sure, but the script is here, it works, and the work is almost done.

**Comment by SoniaZaldana** [comment]
Thanks for the pointers @tstuefe @erikj79. I updated the script to use `\{0,1\}` instead.

**Review by erikj79 — APPROVED** [review]
Looks good to me. I haven't tried it myself though.

File: make/scripts/update_copyright_year.sh [hunk 1]
```diff
@@ -49,66 +51,74 @@ rm -f -r ${tmp}
...
+  echo "to those in the current branch and the current year."
+  echo
+  echo "Note, cancelling the script will skip cleanup in /tmp."
+  echo
+  echo $usage
+  echo "options:"
+  echo "c     Specifies the company. Set to Oracle by default."
+  echo "y     Specifies the copyright year. Set to current year by default."
+  echo "h     Print this help."
+  echo "f     Updates all change sets in a full year."
```
Comment (tstuefe) [inline make/scripts/update_copyright_year.sh:70]: - Please specify the leading dash on all options
- I would print "y" after "f" and write something along these lines "If -f is specified, defines the year of changes for which to modify the copyright (current year if omitted)"

File: make/scripts/update_copyright_year.sh [hunk 2]
```diff
@@ -49,66 +51,74 @@ rm -f -r ${tmp}
...
 
-  if [ "$git_found" == "true" ] && [ "$hg_found" == "false" ] ; then
-    vcs="git"
-  elif [ "$hg_found" == "true" ] && [ "$git_found" == "false" ] ; then
-    vcs="hg"
+# VCS check
+git_found=false
+[ -d "${this_script_dir}/../../.git" ] && git_found=true
+if [ "$git_found" != "true" ]; then
+  echo "Error: could not auto-detect git as the version control system"
```
Comment (tstuefe) [inline make/scripts/update_copyright_year.sh:103]: Pre-existing. The error message is confusing (I first tried your script from a different location). Better would be adding something like this:

"Please execute script from within make/scripts"

**Review by tstuefe — CHANGES_REQUESTED** [review]
A lot better, but not fully there yet. Can you make it so that it also picks up SAP's copyrights when they specify multiple years? They are weird in that the last year does not have a trailing comma, e.g.: 
```
* Copyright (c) 2020, 2023 SAP SE. All rights reserved.
```

File: make/scripts/update_copyright_year.sh (hunk 1 above)
Comment (SoniaZaldana) [reply make/scripts/update_copyright_year.sh:70]: Hi Thomas, 

I assume with this we hope to clarify what time period of changes we are addressing with `-f`. Would something like this communicate what you suggested? 

```
Usage: update_copyright_year.sh [-c company] [-y year] [-h|f]
options:
-c     Specifies the company. Set to Oracle by default.
-y     Specifies the copyright year. Set to current year by default.
-f     Updates the copyright for all change sets in a given year,
       as specified by -y.
-h     Print this help.
```

**Comment by SoniaZaldana** [comment]
> A lot better, but not fully there yet. Can you make it so that it also picks up SAP's copyrights when they specify multiple years? They are weird in that the last year does not have a trailing comma, e.g.:
> 
> ```
> * Copyright (c) 2020, 2023 SAP SE. All rights reserved.
> ```

Nice catch! I also noticed the trailing comma was mistakenly getting added to these cases. I've pushed the fix.

**Review by tstuefe — APPROVED** [review]
Good. Thank you!

**Review by erikj79 — APPROVED** [review]

**Pull request closed without merging**